keps/sig-node/4815-dra-partitionable-devices/README.md

```diff
@@ -0,0 +1,3411 @@
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+Proposed changes:
+```
+// ResourceSliceSpec contains the information published by the driver in one ResourceSlice.
+type ResourceSliceSpec struct {
+		...
+
+		// DeviceMixins represents a list of device mixins, i.e. a collection of
+		// shared attributes and capacities that an actual device can "include"
+		// to extend the set of attributes and capacities it already defines.
+		//
+		// The main purposes of these mixins is to reduce the memory footprint
+		// of devices since they can reference the mixins provided here rather
+		// than duplicate them.
+		//
+		// Must not have more than 128 entries.
+		//
+		// +optional
+		// +listType=atomic
+		DeviceMixins []DeviceMixin
+}
+
+// DeviceMixin defines a specific device mixin for each device type.
+// Besides the name, exactly one field must be set.
+type DeviceMixin struct {
+		// Name is a unique identifier among all mixins managed by the driver
+		// in the pool. It must be a DNS label.
+		//
+		// +required
+		Name string
+
+		// Basic defines a mixin usable by a basic device.
+		//
+		// +optional
+		// +oneOf=deviceMixinType
+		Basic *BasicDeviceMixin
+}
+
+// BasicDeviceMixin defines a mixin that a basic device can include.
+type BasicDeviceMixin struct {
+		// Includes defines the set of other mixins that this mixin includes.
+		//
+		// The propertes of each included mixin are applied to this mixin in
+		// order (recursively). Conflicting properties from multiple mixins are
+		// taken from the last mixin listed that contains them.
+		//
+		// The maximum number of mixins that can be included is 8.
+		//
+		// +optional
+		Includes []DeviceMixinRef
+
+		// Attributes defines the set of attributes for this mixin.
+		// The name of each attribute must be unique in that set.
+		//
+		// The maximum number of attributes and capacities combined is 32.
+		//
+		// +optional
+		Attributes map[QualifiedName]DeviceAttribute
+
+		// Capacity defines the set of capacities for this mixin.
+		// The name of each capacity must be unique in that set.
+		//
+		// The maximum number of attributes and capacities combined is 32.
+		//
+		// +optional
+		Capacity map[QualifiedName]DeviceCapacity
+}
+
+// Device represents one individual hardware instance that can be selected based
+// on its attributes. Besides the name, exactly one field must be set.
+type Device struct {
+		// Name is unique identifier among all devices managed by
+		// the driver in the pool. It must be a DNS label.
+		//
+		// +required
+		Name string
+
+		// Basic defines one device instance.
+		//
+		// +optional
+		// +oneOf=deviceType
+		Basic *BasicDevice
```

Comment:
I don't think there would be any downside. I was under the impression we could always add new fields to an existing type (just not remove them). But I can see how this might break older clients on a downgrade.

Reply:
Let's consider *how* it affects clients which don't know about the new fields.

For `Includes`, the effect will be that a CEL expression which references an attribute defined only in the mixin may trigger a runtime error. Such errors indicate that there's something wrong, which is what we want here.

If the CEL expression is written so that it checks for existence of the attribute, ignoring the `Includes` leads to a different result of the CEL expression. This is worse.

Where it gets really bad (IMHO) is for `ConsumesCapacityFrom`: an old scheduler would ignore that constraint and might allocate two devices which are mutually exclusive.

I think it would be safer to add both fields in a new `PartitionableDevice`. The the failure modes become more favorable: an old client which is looking for devices to allocate won't see those devices and instead of doing something wrong will do nothing.

We can keep the limit on number of devices the same as now, except that we now sum up the length of `BasicDevice`, `PartitionableDevice` *and*  `DeviceMixin` slices when comparing against it.

Note that we cannot lower the limit because making it depend on whether mixins are used would be tricky. Mixins still help in practice by reducing the size of the `ResourceSlice` objects.

Reply:
I am fine with adding a new `PartitionableDevice` type for this. That leaves @johnbelamaric's question from here still open: https://github.com/kubernetes/enhancements/pull/4874#discussion_r1779035937. I am inclined to keep the one-of in the mixin (but only start supporting devices as `PartitonableDevice` with no field for `BasicDevice`).

The alternative *would* have been to have a top-level switch for `BasicDevices` with the device list and its mixins embedded underneath it, but we decided against this top-level switch in favor of a per-device one-of in 1.31.

Reply:
Also note, I'm planning on changing `ConsumesCapacityFrom` to just `consumedCapacity` with the actual capacity being consumed and its source listed under this explicitly. That way a device can *both* consume capacity from a some device *and* declare its own capacity (that another device might want to consume).

Reply:
> I am inclined to keep the one-of in the mixin (but only start supporting devices as PartitonableDevice with no field for BasicDevice).

Makes sense to me.

> Also note, I'm considering changing ConsumesCapacityFrom to just ConsumedCapacity with the actual capacity being consumed

I haven't looked at that part yet.

Reply:
Actually, on second thought adding a separate `ConsumedCapacity` makes it harder to define mixins for capacity because you don't want the mixins to be tied to a particular device instance they are consuming from. I'll leave it as is for now and we can iterate on this once I've added some text describing it.